File: app/src/main/java/org/tvheadend/tvhclient/data/dao/ServerProfileDao.java
package org.tvheadend.tvhclient.data.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import org.tvheadend.tvhclient.data.entity.ServerProfile;

import java.util.List;

@Dao
public interface ServerProfileDao {

    @Query("SELECT p.* FROM server_profiles AS p " +
            "LEFT JOIN connections AS c ON c.id = p.connection_id AND c.active = 1 " +
            "WHERE p.type = 'htsp_playback'")
    List<ServerProfile> loadHtspPlaybackProfilesSync();

    @Query("SELECT p.* FROM server_profiles AS p " +
            "LEFT JOIN connections AS c ON c.id = p.connection_id AND c.active = 1 " +
            "WHERE p.type = 'http_playback'")
    List<ServerProfile> loadHttpPlaybackProfilesSync();

    @Query("SELECT p.* FROM server_profiles AS p " +
            "LEFT JOIN connections AS c ON c.id = p.connection_id AND c.active = 1 " +
            "WHERE p.type = 'recording'")
    List<ServerProfile> loadAllRecordingProfilesSync();

    @Insert
    void insert(ServerProfile serverProfile);

    @Update
    void update(ServerProfile serverProfile);

    @Delete
    void delete(ServerProfile serverProfile);

    @Query("DELETE FROM server_profiles")
    void deleteAll();

    @Query("SELECT p.* FROM server_profiles AS p " +
            "LEFT JOIN connections AS c ON c.id = p.connection_id AND c.active = 1 " +
            "WHERE p.id = :id")
    ServerProfile loadProfileByIdSync(int id);

    @Query("SELECT p.* FROM server_profiles AS p " +
            "LEFT JOIN connections AS c ON c.id = p.connection_id AND c.active = 1 " +
            "WHERE p.uuid = :uuid")
    ServerProfile loadProfileByUuidSync(String uuid);
}

File: app/src/main/java/org/tvheadend/tvhclient/data/dao/ServerProfileDao.kt
package org.tvheadend.tvhclient.data.dao

import androidx.room.*
import org.tvheadend.tvhclient.data.entity.ServerProfile

@Dao
interface ServerProfileDao {

    @get:Query("SELECT COUNT (*) FROM server_profiles AS p " +
            " WHERE $CONNECTION_IS_ACTIVE")
    val itemCountSync: Int

    @Query("SELECT p.* FROM server_profiles AS p " +
            " WHERE $CONNECTION_IS_ACTIVE" +
            " AND p.type = 'htsp_playback'")
    fun loadHtspPlaybackProfilesSync(): List<ServerProfile>

    @Query("SELECT p.* FROM server_profiles AS p " +
            " WHERE $CONNECTION_IS_ACTIVE" +
            " AND p.type = 'http_playback'")
    fun loadHttpPlaybackProfilesSync(): List<ServerProfile>

    @Query("SELECT p.* FROM server_profiles AS p " +
            " WHERE $CONNECTION_IS_ACTIVE" +
            " AND p.type = 'recording'")
    fun loadAllRecordingProfilesSync(): List<ServerProfile>

    @Insert
    fun insert(serverProfile: ServerProfile)

    @Update
    fun update(serverProfile: ServerProfile)

    @Delete
    fun delete(serverProfile: ServerProfile)

    @Query("DELETE FROM server_profiles")
    fun deleteAll()

    @Query("SELECT p.* FROM server_profiles AS p " +
            " WHERE $CONNECTION_IS_ACTIVE" +
            " AND p.id = :id")
    fun loadProfileByIdSync(id: Int): ServerProfile

    @Query("SELECT p.* FROM server_profiles AS p " +
            " WHERE $CONNECTION_IS_ACTIVE" +
            " AND p.uuid = :uuid")
    fun loadProfileByUuidSync(uuid: String): ServerProfile

    companion object {

        const val CONNECTION_IS_ACTIVE = " p.connection_id IN (SELECT id FROM connections WHERE active = 1) "
    }
}
